pkg/services/ngalert/api/api_prometheus.go

```diff
@@ -291,6 +367,22 @@ func (srv PrometheusSrv) toRuleGroup(groupKey ngmodels.AlertRuleGroupKey, folder
 				newRule.Health = "error"
 			}
 
+			if len(withStates) > 0 {
+				if _, ok := withStates[alertState.State]; !ok {
+					continue
+				}
+			}
+
+			if len(matchers) > 0 {
+				lbs := make(model.LabelSet)
+				for k, v := range alertState.Labels {
+					lbs[model.LabelName(k)] = model.LabelValue(v)
+				}
+				if !matchers.Matches(lbs) {
+					continue
+				}
```

Comment:
Method `labels.Matchers.Matches` is very simple 
```
// Matches checks whether all matchers are fulfilled against the given label set.
func (ms Matchers) Matches(lset model.LabelSet) bool {
	for _, m := range ms {
		if !m.Matches(string(lset[model.LabelName(m.Name)])) {
			return false
		}
	}
	return true
}
```
and uses a string. So, if we implement our own `Matchers` that accepts `data.Labels` then we can avoid copying maps.

Reply:
OK, will do! 🙂